pkgs/jnigen/example/kotlin_plugin/lib/kotlin_bindings.dart

```diff
@@ -199,3 +208,246 @@ final class $Example$Type$ extends jni$_.JType<Example> {
+      }
+    } else {
+      $o = $r;
+    }
+    return $o.as<jni$_.JString>(
+      const jni$_.$JString$Type$(),
+      releaseOriginal: true,
+    );
+  }
+
+  /// Maps a specific port to the implemented interface.
+  static final core$_.Map<int, $Thinker> _$impls = {};
+  static jni$_.JObjectPtr _$invoke(
+    int port,
+    jni$_.JObjectPtr descriptor,
+    jni$_.JObjectPtr args,
+  ) {
+    return _$invokeMethod(
+      port,
+      jni$_.MethodInvocation.fromAddresses(
+        0,
+        descriptor.address,
+        args.address,
+      ),
+    );
+  }
+
+  static final jni$_.Pointer<
+          jni$_.NativeFunction<
+              jni$_.JObjectPtr Function(
+                  jni$_.Int64, jni$_.JObjectPtr, jni$_.JObjectPtr)>>
+      _$invokePointer = jni$_.Pointer.fromFunction(_$invoke);
+
+  static jni$_.Pointer<jni$_.Void> _$invokeMethod(
+    int $p,
+    jni$_.MethodInvocation $i,
+  ) {
+    try {
+      final $d = $i.methodDescriptor.toDartString(releaseOriginal: true);
+      final $a = $i.args;
+      if ($d ==
+          r'message(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;') {
+        final $r = _$impls[$p]!.message(
+          $a![0]!.as(const jni$_.$JObject$Type$(), releaseOriginal: true),
+        );
+        return ($r as jni$_.JObject?)
+                ?.as(const jni$_.$JObject$Type$())
+                .reference
+                .toPointer() ??
+            jni$_.nullptr;
+      }
+    } catch (e) {
+      return jni$_.ProtectedJniExtensions.newDartException(e);
+    }
+    return jni$_.nullptr;
+  }
+
+  static void implementIn(
+    jni$_.JImplementer implementer,
+    $Thinker $impl,
+  ) {
+    late final jni$_.RawReceivePort $p;
+    $p = jni$_.RawReceivePort(($m) {
+      if ($m == null) {
+        _$impls.remove($p.sendPort.nativePort);
+        $p.close();
+        return;
+      }
+      final $i = jni$_.MethodInvocation.fromMessage($m);
+      final $r = _$invokeMethod($p.sendPort.nativePort, $i);
+      jni$_.ProtectedJniExtensions.returnResult($i.result, $r);
+    });
+    implementer.add(
+      r'Thinker',
+      $p,
+      _$invokePointer,
+      [],
+    );
+    final $a = $p.sendPort.nativePort;
+    _$impls[$a] = $impl;
+  }
+
+  factory Thinker.implement(
+    $Thinker $impl,
+  ) {
+    final $i = jni$_.JImplementer();
+    implementIn($i, $impl);
+    return Thinker.fromReference(
+      $i.implementReference(),
+    );
+  }
+}
+
+abstract base mixin class $Thinker {
+  factory $Thinker({
+    required core$_.Future<jni$_.JString> Function() message,
+  }) = _$Thinker;
+
+  jni$_.JObject message(jni$_.JObject continuation);
+}
+
+final class _$Thinker with $Thinker {
+  _$Thinker({
+    required core$_.Future<jni$_.JString> Function() message,
+  }) : _message = message;
+
+  final core$_.Future<jni$_.JString> Function() _message;
+
+  jni$_.JObject message(jni$_.JObject continuation) {
+    return jni$_.KotlinContinuation.fromReference(continuation.reference)
```

Comment:
[For the complete PR] We want to enable users to directly mixin `$Thinker` if they choose and not only use the default constructor, so the logic you wrote should be in the `_$invokeMethod` rather than the `_$Thinker`